File: Proline-Databases/src/main/java/fr/proline/context/ContextFactory.java
package fr.proline.context;

import java.sql.SQLException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import fr.proline.repository.IDataStoreConnectorFactory;
import fr.proline.repository.IDatabaseConnector;

public final class ContextFactory {

    private static final Logger LOG = LoggerFactory.getLogger(ContextFactory.class);

    /* Private constructor (Utility class) */
    private ContextFactory() {
    }

    /**
     * Creates an ExecutionContext instance from given DataStoreConnectorFactory and project Id.
     * 
     * @param dsFactory
     *            Factory of Proline DataStore connectors.
     * @param projectId
     *            Id of project to retrieve project specific Dbs (MSI, LCMS).
     * @param useJPA
     *            If <code>true</code> all returned Db contexts wrap a new <code>EntityManager</code> for
     *            relevant Proline Db. If <code>false</code> all returned Db contexts wrap a new SQL JDBC
     *            <code>Connection</code>.
     * @return A new instance of ExecutionContext
     */
    public static IExecutionContext getExecutionContextInstance(final IDataStoreConnectorFactory dsFactory,
	    final int projectId, final boolean useJPA) {
	DatabaseConnectionContext udsDb = null;

	final IDatabaseConnector udsDbConnector = dsFactory.getUdsDbConnector();

	if (udsDbConnector != null) {
	    udsDb = buildDbConnectionContext(udsDbConnector, useJPA);
	}

	DatabaseConnectionContext pdiDb = null;

	final IDatabaseConnector pdiDbConnector = dsFactory.getPdiDbConnector();

	if (pdiDbConnector != null) {
	    pdiDb = buildDbConnectionContext(pdiDbConnector, useJPA);
	}

	DatabaseConnectionContext psDb = null;

	final IDatabaseConnector psDbConnector = dsFactory.getPsDbConnector();

	if (psDbConnector != null) {
	    psDb = buildDbConnectionContext(psDbConnector, useJPA);
	}

	/* Project specific Dbs */
	DatabaseConnectionContext msiDb = null;

	final IDatabaseConnector msiDbConnector = dsFactory.getMsiDbConnector(projectId);

	if (msiDbConnector != null) {
	    msiDb = buildDbConnectionContext(msiDbConnector, useJPA);
	}

	DatabaseConnectionContext lcMsDb = null;

	final IDatabaseConnector lcMsDbConnector = dsFactory.getLcMsDbConnector(projectId);

	if (lcMsDbConnector != null) {
	    lcMsDb = buildDbConnectionContext(lcMsDbConnector, useJPA);
	}

	return new BasicExecutionContext(udsDb, pdiDb, psDb, msiDb, lcMsDb);
    }

    /**
     * Creates a <code>DatabaseConnectionContext</code> from given DatabaseConnector.
     * 
     * @param dbConnector
     *            DatabaseConnector used to access a Proline Db.
     * @param useJPA
     *            If <code>true</code> returned Db context wraps a new <code>EntityManager</code>. If
     *            <code>false</code> returned Db context wraps a new SQL JDBC <code>Connection</code>.
     * @return A new instance of <code>DatabaseConnectionContext</code>
     */
    public static DatabaseConnectionContext buildDbConnectionContext(final IDatabaseConnector dbConnector,
	    final boolean useJPA) {

	if (dbConnector == null) {
	    throw new IllegalArgumentException("DbConnector is null");
	}

	DatabaseConnectionContext dbConnectionContext = null;

	if (useJPA) {
	    dbConnectionContext = new DatabaseConnectionContext(dbConnector);
	} else {

	    try {
		dbConnectionContext = new DatabaseConnectionContext(dbConnector.getDataSource()
			.getConnection(), dbConnector.getDriverType());
	    } catch (SQLException ex) {
		/* Log and re-throw */
		final String message = "Unable to obtain SQL JDBC Connection for "
			+ dbConnector.getProlineDatabaseType();
		LOG.error(message, ex);

		throw new RuntimeException(message, ex);
	    }

	} // End if (SQL mode)

	return dbConnectionContext;
    }

}

File: Proline-DAL/src/main/scala/fr/proline/core/dal/ContextFactory.scala
package fr.proline.core.dal

import java.sql.SQLException

import com.typesafe.scalalogging.LazyLogging
import fr.profi.util.ThreadLogger
import fr.proline.context._
import fr.proline.core.om.provider.PeptideCacheExecutionContext
import fr.proline.repository._

object BuildLazyExecutionContext extends LazyLogging {

  /**
   * Creates a LazyExecutionContext instance from given IDataStoreConnectorFactory and project Id.
   *
   * @param dsFactory
   *            Factory of Proline DataStore connectors.
   * @param projectId
   *            Id of project to retrieve project specific Dbs (MSI, LCMS).
   * @param useJPA
   *            If <code>true</code> all returned Db contexts wrap a new <code>EntityManager</code> for
   *            relevant Proline Db. If <code>false</code> all returned Db contexts wrap a new SQL JDBC
   *            <code>Connection</code>.
   * @return A new instance of ExecutionContext
   */
  def apply(
    dsFactory: IDataStoreConnectorFactory,
    projectId: Long,
    useJPA: Boolean,
    onConnectionContextClose: Option[IDatabaseConnector => Unit] = None
  ): IExecutionContext = {
    val currentThread = Thread.currentThread

    if (!currentThread.getUncaughtExceptionHandler.isInstanceOf[ThreadLogger]) {
      currentThread.setUncaughtExceptionHandler(new ThreadLogger(logger.underlying.getName()))
    }

    val udsDbCtxBuilder = { () =>
      val udsDbConnector = dsFactory.getUdsDbConnector
      if (udsDbConnector == null) null
      else BuildUdsDbConnectionContext(udsDbConnector, useJPA, onConnectionContextClose)
    }
    
    /* Project specific Dbs */
    val msiDbCtxBuilder = { () =>
      val msiDbConnector = dsFactory.getMsiDbConnector(projectId)
      if (msiDbConnector == null) null
      else BuildMsiDbConnectionContext(msiDbConnector, useJPA, onConnectionContextClose)
    }
    
    val lcMsDbCtxBuilder = { () =>
      val lcMsDbConnector = dsFactory.getLcMsDbConnector(projectId)
      if (lcMsDbConnector == null) null
      else BuildLcMsDbConnectionContext(lcMsDbConnector, useJPA, onConnectionContextClose)
    }

    val lazyContext = new LazyExecutionContext(
      projectId,
      useJPA,
      udsDbCtxBuilder,
      msiDbCtxBuilder,
      lcMsDbCtxBuilder
    )

    new PeptideCacheExecutionContext(lazyContext)
  }

}


object BuildDbConnectionContext extends LazyLogging {
  
  /**
   * Creates a <code>DatabaseConnectionContext</code> from given DatabaseConnector.
   *
   * @param dbConnector
   *            DatabaseConnector used to access a Proline Db.
   * @param useJPA
   *            If <code>true</code> returned Db context wraps a new <code>EntityManager</code>. If
   *            <code>false</code> returned SQLConnectionContext wraps a new SQL JDBC <code>Connection</code>.
   * @return A new instance of <code>DatabaseConnectionContext</code> or <code>SQLConnectionContext</code> for SQL
   */
  def apply(
    dbConnector: IDatabaseConnector,
    useJPA: Boolean,
    onClose: Option[IDatabaseConnector => Unit] = None
  ): DatabaseConnectionContext = {
    
    if (dbConnector == null) {
      throw new IllegalArgumentException("DbConnector is null")
    }
    
    val dbType = dbConnector.getProlineDatabaseType

    val dbCtx = if (useJPA) {
      dbType match {
        case ProlineDatabaseType.LCMS => new LcMsDbConnectionContext(dbConnector)
        case ProlineDatabaseType.MSI => new MsiDbConnectionContext(dbConnector)
        case ProlineDatabaseType.UDS => new UdsDbConnectionContext(dbConnector)
        case _ => new DatabaseConnectionContext(dbConnector)
      }
    } else {

      try {
        val connection = dbConnector.getDataSource.getConnection
        
        dbType match {
          case ProlineDatabaseType.LCMS => new LcMsDbConnectionContext(connection, dbConnector.getDriverType)
          case ProlineDatabaseType.MSI => new MsiDbConnectionContext(connection, dbConnector.getDriverType)
          case ProlineDatabaseType.UDS => new UdsDbConnectionContext(connection, dbConnector.getDriverType)
          case _ => new DatabaseConnectionContext(connection, dbType, dbConnector.getDriverType)
        }
        
      } catch {

        case sqlEx: SQLException => {
          /* Log and re-throw */
          val message = "Unable to obtain SQL JDBC Connection for " + dbConnector.getProlineDatabaseType
          logger.error(message, sqlEx)

          throw new RuntimeException(message, sqlEx)
        }

      }

    } // End if (SQL mode)

    // Set onCloseCallback if provided
    dbCtx.setOnCloseCallback(new Runnable {
      def run() {
        if (onClose.isDefined) onClose.get.apply(dbConnector)
      }
    })
    
    dbCtx
  }

}

object BuildLcMsDbConnectionContext extends LazyLogging {

  def apply(dbConnector: IDatabaseConnector, useJPA: Boolean, onClose: Option[IDatabaseConnector => Unit] = None): LcMsDbConnectionContext = {
    require( dbConnector.getProlineDatabaseType == ProlineDatabaseType.LCMS, "wrong ProlineDatabaseType")
    
    BuildDbConnectionContext(dbConnector, useJPA, onClose).asInstanceOf[LcMsDbConnectionContext]
  }

}

object BuildMsiDbConnectionContext extends LazyLogging {

  def apply(dbConnector: IDatabaseConnector, useJPA: Boolean, onClose: Option[IDatabaseConnector => Unit] = None): MsiDbConnectionContext = {
    require( dbConnector.getProlineDatabaseType == ProlineDatabaseType.MSI, "wrong ProlineDatabaseType")
    
    BuildDbConnectionContext(dbConnector, useJPA, onClose).asInstanceOf[MsiDbConnectionContext]
  }
}



object BuildUdsDbConnectionContext extends LazyLogging {

  def apply(dbConnector: IDatabaseConnector, useJPA: Boolean, onClose: Option[IDatabaseConnector => Unit] = None): UdsDbConnectionContext = {
    require( dbConnector.getProlineDatabaseType == ProlineDatabaseType.UDS, "wrong ProlineDatabaseType")
    
    BuildDbConnectionContext(dbConnector, useJPA, onClose).asInstanceOf[UdsDbConnectionContext]
  }

}

